File: backend/feature_flags.py
"""
Feature Flags System
Allows dynamic toggling of application modules and features
"""
from typing import Dict, Any
import json
from pathlib import Path
from datetime import datetime
import logging

logger = logging.getLogger(__name__)


class FeatureFlagManager:
    """Manage application feature flags"""

    DEFAULT_FLAGS = {
        "enableWhaleTracking": True,
        "enableMarketOverview": True,
        "enableFearGreedIndex": True,
        "enableNewsFeed": True,
        "enableSentimentAnalysis": True,
        "enableMlPredictions": False,  # Disabled by default (requires HF setup)
        "enableProxyAutoMode": True,
        "enableDefiProtocols": True,
        "enableTrendingCoins": True,
        "enableGlobalStats": True,
        "enableProviderRotation": True,
        "enableWebSocketStreaming": True,
        "enableDatabaseLogging": True,
        "enableRealTimeAlerts": False,  # New feature - not yet implemented
        "enableAdvancedCharts": True,
        "enableExportFeatures": True,
        "enableCustomProviders": True,
        "enablePoolManagement": True,
        "enableHFIntegration": True,
    }

    def __init__(self, storage_path: str = "data/feature_flags.json"):
        """
        Initialize feature flag manager

        Args:
            storage_path: Path to persist feature flags
        """
        self.storage_path = Path(storage_path)
        self.flags = self.DEFAULT_FLAGS.copy()
        self.load_flags()

    def load_flags(self):
        """Load feature flags from storage"""
        try:
            if self.storage_path.exists():
                with open(self.storage_path, 'r', encoding='utf-8') as f:
                    saved_flags = json.load(f)
                # Merge saved flags with defaults (in case new flags were added)
                self.flags.update(saved_flags.get('flags', {}))
                logger.info(f"Loaded feature flags from {self.storage_path}")
            else:
                # Create storage directory if it doesn't exist
                self.storage_path.parent.mkdir(parents=True, exist_ok=True)
                self.save_flags()
                logger.info("Initialized default feature flags")
        except Exception as e:
            logger.error(f"Error loading feature flags: {e}")
            self.flags = self.DEFAULT_FLAGS.copy()

    def save_flags(self):
        """Save feature flags to storage"""
        try:
            self.storage_path.parent.mkdir(parents=True, exist_ok=True)
            data = {
                'flags': self.flags,
                'last_updated': datetime.now().isoformat()
            }
            with open(self.storage_path, 'w', encoding='utf-8') as f:
                json.dump(data, f, indent=2)
            logger.info("Feature flags saved successfully")
        except Exception as e:
            logger.error(f"Error saving feature flags: {e}")

    def get_all_flags(self) -> Dict[str, bool]:
        """Get all feature flags"""
        return self.flags.copy()

    def get_flag(self, flag_name: str) -> bool:
        """
        Get a specific feature flag value

        Args:
            flag_name: Name of the flag

        Returns:
            bool: Flag value (defaults to False if not found)
        """
        return self.flags.get(flag_name, False)

    def set_flag(self, flag_name: str, value: bool) -> bool:
        """
        Set a feature flag value

        Args:
            flag_name: Name of the flag
            value: New value (True/False)

        Returns:
            bool: Success status
        """
        try:
            self.flags[flag_name] = bool(value)
            self.save_flags()
            logger.info(f"Feature flag '{flag_name}' set to {value}")
            return True
        except Exception as e:
            logger.error(f"Error setting feature flag: {e}")
            return False

    def update_flags(self, updates: Dict[str, bool]) -> bool:
        """
        Update multiple flags at once

        Args:
            updates: Dictionary of flag name -> value pairs

        Returns:
            bool: Success status
        """
        try:
            for flag_name, value in updates.items():
                self.flags[flag_name] = bool(value)
            self.save_flags()
            logger.info(f"Updated {len(updates)} feature flags")
            return True
        except Exception as e:
            logger.error(f"Error updating feature flags: {e}")
            return False

    def reset_to_defaults(self) -> bool:
        """Reset all flags to default values"""
        try:
            self.flags = self.DEFAULT_FLAGS.copy()
            self.save_flags()
            logger.info("Feature flags reset to defaults")
            return True
        except Exception as e:
            logger.error(f"Error resetting feature flags: {e}")
            return False

    def is_enabled(self, flag_name: str) -> bool:
        """
        Check if a feature is enabled (alias for get_flag)

        Args:
            flag_name: Name of the flag

        Returns:
            bool: True if enabled, False otherwise
        """
        return self.get_flag(flag_name)

    def get_enabled_features(self) -> Dict[str, bool]:
        """Get only enabled features"""
        return {k: v for k, v in self.flags.items() if v is True}

    def get_disabled_features(self) -> Dict[str, bool]:
        """Get only disabled features"""
        return {k: v for k, v in self.flags.items() if v is False}

    def get_flag_count(self) -> Dict[str, int]:
        """Get count of enabled/disabled flags"""
        enabled = sum(1 for v in self.flags.values() if v)
        disabled = len(self.flags) - enabled
        return {
            'total': len(self.flags),
            'enabled': enabled,
            'disabled': disabled
        }

    def get_feature_info(self) -> Dict[str, Any]:
        """Get comprehensive feature flag information"""
        counts = self.get_flag_count()
        return {
            'flags': self.flags,
            'counts': counts,
            'enabled_features': list(self.get_enabled_features().keys()),
            'disabled_features': list(self.get_disabled_features().keys()),
            'storage_path': str(self.storage_path),
            'last_loaded': datetime.now().isoformat()
        }


# Global instance
feature_flags = FeatureFlagManager()


# Convenience functions
def is_feature_enabled(flag_name: str) -> bool:
    """Check if a feature is enabled"""
    return feature_flags.is_enabled(flag_name)


def get_all_feature_flags() -> Dict[str, bool]:
    """Get all feature flags"""
    return feature_flags.get_all_flags()


def set_feature_flag(flag_name: str, value: bool) -> bool:
    """Set a feature flag"""
    return feature_flags.set_flag(flag_name, value)


def update_feature_flags(updates: Dict[str, bool]) -> bool:
    """Update multiple feature flags"""
    return feature_flags.update_flags(updates)

File: static/js/feature-flags.js
/**
 * Feature Flags Manager - Frontend
 * Handles feature flag state and synchronization with backend
 */

class FeatureFlagsManager {
    constructor() {
        this.flags = {};
        this.localStorageKey = 'crypto_monitor_feature_flags';
        this.apiEndpoint = '/api/feature-flags';
        this.listeners = [];
    }

    /**
     * Initialize feature flags from backend and localStorage
     */
    async init() {
        // Load from localStorage first (for offline/fast access)
        this.loadFromLocalStorage();

        // Sync with backend
        await this.syncWithBackend();

        // Set up periodic sync (every 30 seconds)
        setInterval(() => this.syncWithBackend(), 30000);

        return this.flags;
    }

    /**
     * Load flags from localStorage
     */
    loadFromLocalStorage() {
        try {
            const stored = localStorage.getItem(this.localStorageKey);
            if (stored) {
                const data = JSON.parse(stored);
                this.flags = data.flags || {};
                console.log('[FeatureFlags] Loaded from localStorage:', this.flags);
            }
        } catch (error) {
            console.error('[FeatureFlags] Error loading from localStorage:', error);
        }
    }

    /**
     * Save flags to localStorage
     */
    saveToLocalStorage() {
        try {
            const data = {
                flags: this.flags,
                updated_at: new Date().toISOString()
            };
            localStorage.setItem(this.localStorageKey, JSON.stringify(data));
            console.log('[FeatureFlags] Saved to localStorage');
        } catch (error) {
            console.error('[FeatureFlags] Error saving to localStorage:', error);
        }
    }

    /**
     * Sync with backend
     */
    async syncWithBackend() {
        try {
            const response = await fetch(this.apiEndpoint);
            if (!response.ok) {
                throw new Error(`HTTP ${response.status}`);
            }

            const data = await response.json();
            this.flags = data.flags || {};
            this.saveToLocalStorage();
            this.notifyListeners();

            console.log('[FeatureFlags] Synced with backend:', this.flags);
            return this.flags;
        } catch (error) {
            console.error('[FeatureFlags] Error syncing with backend:', error);
            // Fall back to localStorage
            return this.flags;
        }
    }

    /**
     * Check if a feature is enabled
     */
    isEnabled(flagName) {
        return this.flags[flagName] === true;
    }

    /**
     * Get all flags
     */
    getAll() {
        return { ...this.flags };
    }

    /**
     * Set a single flag
     */
    async setFlag(flagName, value) {
        try {
            const response = await fetch(`${this.apiEndpoint}/${flagName}`, {
                method: 'PUT',
                headers: {
                    'Content-Type': 'application/json',
                },
                body: JSON.stringify({
                    flag_name: flagName,
                    value: value
                })
            });

            if (!response.ok) {
                throw new Error(`HTTP ${response.status}`);
            }

            const data = await response.json();
            if (data.success) {
                this.flags[flagName] = value;
                this.saveToLocalStorage();
                this.notifyListeners();
                console.log(`[FeatureFlags] Set ${flagName} = ${value}`);
                return true;
            }

            return false;
        } catch (error) {
            console.error(`[FeatureFlags] Error setting flag ${flagName}:`, error);
            return false;
        }
    }

    /**
     * Update multiple flags
     */
    async updateFlags(updates) {
        try {
            const response = await fetch(this.apiEndpoint, {
                method: 'PUT',
                headers: {
                    'Content-Type': 'application/json',
                },
                body: JSON.stringify({
                    flags: updates
                })
            });

            if (!response.ok) {
                throw new Error(`HTTP ${response.status}`);
            }

            const data = await response.json();
            if (data.success) {
                this.flags = data.flags;
                this.saveToLocalStorage();
                this.notifyListeners();
                console.log('[FeatureFlags] Updated flags:', updates);
                return true;
            }

            return false;
        } catch (error) {
            console.error('[FeatureFlags] Error updating flags:', error);
            return false;
        }
    }

    /**
     * Reset to defaults
     */
    async resetToDefaults() {
        try {
            const response = await fetch(`${this.apiEndpoint}/reset`, {
                method: 'POST'
            });

            if (!response.ok) {
                throw new Error(`HTTP ${response.status}`);
            }

            const data = await response.json();
            if (data.success) {
                this.flags = data.flags;
                this.saveToLocalStorage();
                this.notifyListeners();
                console.log('[FeatureFlags] Reset to defaults');
                return true;
            }

            return false;
        } catch (error) {
            console.error('[FeatureFlags] Error resetting flags:', error);
            return false;
        }
    }

    /**
     * Add change listener
     */
    onChange(callback) {
        this.listeners.push(callback);
        return () => {
            const index = this.listeners.indexOf(callback);
            if (index > -1) {
                this.listeners.splice(index, 1);
            }
        };
    }

    /**
     * Notify all listeners of changes
     */
    notifyListeners() {
        this.listeners.forEach(callback => {
            try {
                callback(this.flags);
            } catch (error) {
                console.error('[FeatureFlags] Error in listener:', error);
            }
        });
    }

    /**
     * Render feature flags UI
     */
    renderUI(containerId) {
        const container = document.getElementById(containerId);
        if (!container) {
            console.error(`[FeatureFlags] Container #${containerId} not found`);
            return;
        }

        const flagDescriptions = {
            enableWhaleTracking: 'Show whale transaction tracking',
            enableMarketOverview: 'Display market overview dashboard',
            enableFearGreedIndex: 'Show Fear & Greed sentiment index',
            enableNewsFeed: 'Display cryptocurrency news feed',
            enableSentimentAnalysis: 'Enable sentiment analysis features',
            enableMlPredictions: 'Show ML-powered price predictions',
            enableProxyAutoMode: 'Automatic proxy for failing APIs',
            enableDefiProtocols: 'Display DeFi protocol data',
            enableTrendingCoins: 'Show trending cryptocurrencies',
            enableGlobalStats: 'Display global market statistics',
            enableProviderRotation: 'Enable provider rotation system',
            enableWebSocketStreaming: 'Real-time WebSocket updates',
            enableDatabaseLogging: 'Log provider health to database',
            enableRealTimeAlerts: 'Show real-time alert notifications',
            enableAdvancedCharts: 'Display advanced charting',
            enableExportFeatures: 'Enable data export functions',
            enableCustomProviders: 'Allow custom API providers',
            enablePoolManagement: 'Enable provider pool management',
            enableHFIntegration: 'HuggingFace model integration'
        };

        let html = '<div class="feature-flags-container">';
        html += '<h3>Feature Flags</h3>';
        html += '<div class="feature-flags-list">';

        Object.keys(this.flags).forEach(flagName => {
            const enabled = this.flags[flagName];
            const description = flagDescriptions[flagName] || flagName;

            html += `
                <div class="feature-flag-item">
                    <label class="feature-flag-label">
                        <input
                            type="checkbox"
                            class="feature-flag-toggle"
                            data-flag="${flagName}"
                            ${enabled ? 'checked' : ''}
                        />
                        <span class="feature-flag-name">${description}</span>
                    </label>
                    <span class="feature-flag-status ${enabled ? 'enabled' : 'disabled'}">
                        ${enabled ? '✓ Enabled' : '✗ Disabled'}
                    </span>
                </div>
            `;
        });

        html += '</div>';
        html += '<div class="feature-flags-actions">';
        html += '<button id="ff-reset-btn" class="btn btn-secondary">Reset to Defaults</button>';
        html += '</div>';
        html += '</div>';

        container.innerHTML = html;

        // Add event listeners
        container.querySelectorAll('.feature-flag-toggle').forEach(toggle => {
            toggle.addEventListener('change', async (e) => {
                const flagName = e.target.dataset.flag;
                const value = e.target.checked;
                await this.setFlag(flagName, value);
            });
        });

        const resetBtn = container.querySelector('#ff-reset-btn');
        if (resetBtn) {
            resetBtn.addEventListener('click', async () => {
                if (confirm('Reset all feature flags to defaults?')) {
                    await this.resetToDefaults();
                    this.renderUI(containerId);
                }
            });
        }

        // Listen for changes and re-render
        this.onChange(() => {
            this.renderUI(containerId);
        });
    }
}

// Global instance
window.featureFlagsManager = new FeatureFlagsManager();

// Auto-initialize on DOMContentLoaded
document.addEventListener('DOMContentLoaded', () => {
    window.featureFlagsManager.init().then(() => {
        console.log('[FeatureFlags] Initialized');
    });
});
